### PR TITLE
Add systemctl list-paths

### LINKED ISSUE #6054: [RFE] systemctl list-paths
### Submission type

 - Request for enhancement (RFE)

systemctl currently supports `list-{units,sockets,timers}`

The addition of `systemctl list-paths` would help simplify reviewing existing path watchers.

### systemd version the issue has been seen with

> 233

### Used distribution

> Fedora

### PR DESCRIPTION
**Pull request by dtardon**
Fixes #6054.

### CODE REVIEW

File: src/systemctl/systemctl-list-units.c [hunk 1]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+        int r;
+
+        assert(a);
+        assert(b);
+
+        r = strcasecmp_ptr(a->machine, b->machine);
+        if (r != 0)
+                return r;
+
+        r = strcmp(a->path, b->path);
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: Missing returned value handling.

File: src/systemctl/systemctl-list-units.c [hunk 2]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+        if (!table)
+                return log_oom();
+
+        table_set_header(table, arg_legend != 0);
+        if (arg_full)
+                table_set_width(table, 0);
+
+        table_set_ersatz_string(table, TABLE_ERSATZ_DASH);
+
+        for (struct path_info *p = ps->items; p < ps->items + ps->count; p++) {
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c:1044]: IIRC, writing a loop in such way may trigger warning by Coverity, as we may compare or add with `NULL`.
I'd like to write it by `for (size_t i = 0; i < ps->count; i++)`.

File: src/systemctl/systemctl-list-units.c [hunk 3]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+                _cleanup_free_ char *j = NULL;
+                const char *unit;
+
+                if (p->machine) {
+                        j = strjoin(p->machine, ":", p->id);
+                        if (!j)
+                                return log_oom();
+                        unit = j;
+                } else
+                        unit = p->id;
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: Not a matter, but you can use `table_add_cell_stringf()`.

File: src/systemctl/systemctl-list-units.c [hunk 4]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+
+                if (strv_isempty(p->triggered))
+                        r = table_add_cell(table, NULL, TABLE_EMPTY, NULL);
+                else if (strv_length(p->triggered) == 1)
+                        r = table_add_cell(table, NULL, TABLE_STRING, p->triggered[0]);
+                else
+                        /* This should never happen, currently our path units can only trigger a
+                         * single unit. But let's handle this anyway, who knows what the future
+                         * brings? */
+                        r = table_add_cell(table, NULL, TABLE_STRV, p->triggered);
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: Why not to always use `TABLE_STRV` ??

File: src/systemctl/systemctl-list-units.c [hunk 5]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+        r = strcmp(a->path, b->path);
+
+        r = strcmp(a->condition, b->condition);
+        if (r != 0)
+                return r;
+
+        return strcasecmp_ptr(a->id, b->id);
+}
+
+static void path_infos_free(struct path_infos *ps) {
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: This does not free `struct path_infos`, but frees the list. Please rename this `path_infos_done()`.

File: src/systemctl/systemctl-list-units.c [hunk 6]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+        r = expand_unit_names(bus, strv_skip(argv, 1), ".path", &units, NULL);
+        if (r < 0)
+                return r;
+
+        if (argc == 1 || units) {
+                n = get_unit_list_recursive(bus, units, &unit_infos, &replies, &machines);
+                if (n < 0)
+                        return n;
+
+                for (const UnitInfo *u = unit_infos; u < unit_infos + n; u++) {
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c:1096]: Also, I'd like to use `for (size_t j = 0; j < (size_t) n; j++)`.

File: src/systemctl/systemctl-list-units.c [hunk 7]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+                                        .machine = u->machine,
+                                        .id = u->id,
+                                        .condition = conditions[i],
+                                        .path = paths[i],
+                                        .triggered = triggered,
+                                        .own_triggered = i == 0,
+                                };
+
+                        conditions = mfree(conditions);
+                        paths = mfree(paths);
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: These line sound hackish. Please add a brief comment.

File: src/systemctl/systemctl-list-units.c [hunk 8]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+
+                        for (int i = 0; i < c; i++)
+                                path_infos.items[path_infos.count + i] = (struct path_info) {
+                                        .machine = u->machine,
+                                        .id = u->id,
+                                        .condition = conditions[i],
+                                        .path = paths[i],
+                                        .triggered = triggered,
+                                        .own_triggered = i == 0,
+                                };
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: Most entries are duplicated, how about to simply take `conditions` and `paths` as is, instead of creating the struct with each condition and path?

**Review by yuwata — CHANGES_REQUESTED** [review]
Please add a brief test for the new command in test/units/testsuite-26.sh.
Also, please update the shell completions.

File: src/systemctl/systemctl-list-units.c (hunk 1 above)
Comment (poettering) [reply src/systemctl/systemctl-list-units.c]: please use path_compare()

File: src/systemctl/systemctl-list-units.c [hunk 9]
```diff
@@ -932,3 +932,230 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+
+                        conditions = mfree(conditions);
+                        paths = mfree(paths);
+                        triggered = NULL;
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c]: hmm, you invalidate `triggered` here after adding it to all item objects you just created, hence it's now owned by which code specifically? i.e. it should have one owner who is responsible for freeing it.

Comment (poettering) [reply src/systemctl/systemctl-list-units.c]: oh, now i get it, "own_triggered" indicates ownership. i see.

but hwat happens if c is 0?

Comment (poettering) [reply src/systemctl/systemctl-list-units.c]: please find a different solution (copying with strdup()?), this is too "magic" for my taste

**Comment by poettering** [comment]
looks pretty ok, but the memory management is a bit too adventurous for my taste. just use strdup() instead of maintaining thta ownership boolean. this is not performance critical

File: src/systemctl/systemctl-list-units.c (hunk 2 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:1044]: That should be covered by `assert(ps->items || ps->count > 0);` above. The same pattern is used in other `output_*_list` functions in the file and it doesn't trigger any warnings. But I don't really care; if you think looping over an index is better, I can change it.

File: src/systemctl/systemctl-list-units.c (hunk 4 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: It matters for JSON output, see 50098d87fbe677bd43cb7329b2e32b039401f35c.

File: src/systemctl/systemctl-list-units.c (hunk 5 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: Yeah. I took `sd_bus_error_free` for inspiration, but that name's probably a relic of history that cannot be changed because it's API...

File: src/systemctl/systemctl-list-units.c (hunk 8 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: There are two reasons for this:

1. The function potentially handles multiple units (the outer `for` loop), hence there's no single `paths` (or `conditions`) list there that could just be passed along.
2. The `path_infos` list is sorted at the end and both `.path` and `.condition` are among the criterions. Having them in separate arrays would only complicate things.

File: src/systemctl/systemctl-list-units.c (hunk 9 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: I copied this from `verb_list_sockets` :-) (Granted, that code probably hasn't changed since it was written 10+ years ago. And the coding style has changed since these days, esp. around memory management...) I'll see what I can do to make it more palatable.

File: src/systemctl/systemctl-list-units.c [hunk 10]
```diff
@@ -138,9 +144,9 @@ static int output_units_list(const UnitInfo *unit_infos, size_t c) {
         table_set_ersatz_string(table, TABLE_ERSATZ_DASH);
 
         for (const UnitInfo *u = unit_infos; unit_infos && (size_t) (u - unit_infos) < c; u++) {
-                _cleanup_free_ char *j = NULL;
+                _cleanup_free_ char *id = NULL;
                 const char *on_underline = "", *on_loaded = "", *on_active = "";
-                const char *on_circle = "", *id;
+                const char *on_circle = "";
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: If you touch the lines, then please merge them into one line.

File: src/systemctl/systemctl-list-units.c (hunk 2 above)
Comment (yuwata) [reply src/systemctl/systemctl-list-units.c:1044]: Indeed. Let's keep this as is. Sorry for the noise.

File: src/systemctl/systemctl-list-units.c (hunk 8 above)
Comment (yuwata) [reply src/systemctl/systemctl-list-units.c]: Indeed. That makes sense.

File: src/systemctl/systemctl-list-units.c (hunk 7 above)
Comment (yuwata) [reply src/systemctl/systemctl-list-units.c]: Please add comment.

Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: Sorry, forgot about this one... And also about Lennart's complains about the `.own_triggered` stuff. Resetting the flags as there'll be another update.

File: src/systemctl/systemctl-list-units.c [hunk 11]
```diff
@@ -923,3 +924,215 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+                return log_error_errno(r, "Failed to get paths: %s", bus_error_message(&error, r));
+
+        r = sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "(ss)");
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        while ((r = sd_bus_message_read(reply, "(ss)", &condition, &path)) > 0) {
+                r = strv_extend(&conditions, condition);
+                if (r < 0)
+                        return log_oom();
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c:1007]: you could shorten this to:

```c
if (strv_extend(…) < 0)
        return log_oom();
```

given that strv_extend() only fails with OOM and you hence don#t need to care about the precise return value, only whether it is negative

File: src/systemctl/systemctl-list-units.c [hunk 12]
```diff
@@ -923,3 +924,215 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+
+                r = strv_extend(&paths, path);
+                if (r < 0)
+                        return log_oom();
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c:1010]: similar

File: src/systemctl/systemctl-list-units.c [hunk 13]
```diff
@@ -923,3 +924,215 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+                        if (!GREEDY_REALLOC(path_infos.items, path_infos.count + c))
+                                return log_oom();
+
+                        for (int i = 0; i < c; i++)
+                                path_infos.items[path_infos.count + i] = (struct path_info) {
+                                        .machine = u->machine,
+                                        .id = u->id,
+                                        .condition = conditions[i],
+                                        .path = paths[i],
+                                        .triggered = strv_copy(triggered),
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c]: oom check

File: src/systemctl/systemctl-list-units.c (hunk 11 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:1007]: Done.

File: src/systemctl/systemctl-list-units.c (hunk 12 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:1010]: Done.

File: src/systemctl/systemctl-list-units.c (hunk 13 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: Done. The loop now goes through the strvs (`paths` and `conditions`) backwards, which made the memory handling a bit simpler.

File: src/systemctl/systemctl-list-units.c [hunk 14]
```diff
@@ -922,3 +923,213 @@ int verb_list_automounts(int argc, char *argv[], void *userdata) {
...
+
+        while ((r = sd_bus_message_read(reply, "(ss)", &condition, &path)) > 0) {
+                if (strv_extend(&conditions, condition) < 0)
+                        return log_oom();
+
+                if (strv_extend(&paths, path) < 0)
+                        return log_oom();
+
+                n++;
+        }
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c:1013]: after leaving the loop you should still check if `r` is an error, otherwise we'll miss this error condition of the sd_bus_mesage_read() call in the while expression. i.e. immediately after the `}` just add:

```c
}
if (r < 0)
        return bus_log_parse_error(r);
```

**Comment by poettering** [comment]
one last thing

File: src/systemctl/systemctl-list-units.c (hunk 14 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:1013]: Done.

**Review by yuwata — APPROVED** [review]
LGTM.

**Comment by yuwata** [comment]
All CI failures seem to be unrelated. Merging.
